### PR TITLE
Improved handling for invalid date time. (#31nfzxk)

### LINKED ISSUE #76: Improved handling for invalid date time
## What is the motivation for adding/enhancing this feature?
When user uploads a CSV with an invalid date time format we need a streamlined way of helping them correct their data.

## What are the acceptance criteria?
Clicking on the chip with an invalid date time should open a modal where the user can configure the format they want to use and see why their input is invalid.

## Additional information
Configure date time

### PR DESCRIPTION
**Pull request by k2maan**
### Related Issues

Closes #76 

### Short Description and Why It's Useful

Provides a better way for the user to check if their given arrival date is valid or invalid. In case it is invalid, the user can see what's wrong, what is the accepted format, and change the format based on their needs.

### Screenshots of Visual Changes before/after (If There Are Any)
Import - HotWax Commerce

**IMPORTANT NOTICE** - Remember to add changelog entry

### Contribution and Currently Important Rules Acceptance

- [ ] I read and followed [contribution rules](https://github.com/hotwax/import#contribution-guideline)

### CODE REVIEW

File: src/components/DateTimeConfigModal.vue [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        <p>{{ $t("Expected input") }}</p>
+      </ion-label>
+      <ion-label slot="end">{{ currentDate }}</ion-label>
+    </ion-item>
+    <ion-item lines="none">
+      <ion-label>{{ $t("Custom date format") }}</ion-label>
+      <ion-input clear-input='true' @keyup.enter="dateTimeFormat = $event.target.value;" v-model="dateTimeFormat" :value="dateTimeFormat" :placeholder="defaultDateTimeFormat" />
+    </ion-item>
+
+    <hr />
```
Comment (disha1202) [inline src/components/DateTimeConfigModal.vue:32]: Use lines="full" instead of hr tag

File: src/components/DateTimeConfigModal.vue [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+      </ion-buttons>
+      <ion-title>{{ $t("Configure date time format") }}</ion-title>
+    </ion-toolbar>
+  </ion-header>
+
+  <ion-content class="ion-padding">
+    
+    <ion-item>
+      <ion-label>
+        <p>{{ $t("CSV input") }}</p>
```
Comment (disha1202) [inline src/components/DateTimeConfigModal.vue:17]: Remove all the <p> tags. instead use color="medium"

File: src/views/OrderDetail.vue [hunk 3]
```diff
@@ -122,9 +122,9 @@
               <ion-label>{{ item.quantityOrdered }} {{ $t("Ordered") }}</ion-label>
             </ion-chip>
 
-            <ion-chip outline class="tablet">
+            <ion-chip outline class="tablet" @click="configureFormat(item.arrivalDate)">
               <ion-icon :icon="sendOutline" />
-              <ion-label>{{ item.arrivalDate }}</ion-label>
+              <ion-label>{{ showInvalidDateTime(item.arrivalDate) }}</ion-label>
```
Comment (disha1202) [inline src/views/OrderDetail.vue:127]: Incorrect method name

File: src/components/DateTimeConfigModal.vue [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+    <ion-item class="ion-padding-bottom" lines="none">
+      <p><a href="https://moment.github.io/luxon/#/formatting?id=table-of-tokens" target="_blank">{{ $t('View all date time formats supported by the HotWax Import app.') }}</a></p>
+    </ion-item>
+          
+    <ion-item>
+      <ion-label>
+        <p>{{ $t("Parsed output") }}</p>
+      </ion-label>
+      <ion-label slot="end">{{ isDateInvalid(arrivalDate) ? $t("Invalid input") : arrivalDate }}</ion-label>
```
Comment (disha1202) [inline src/components/DateTimeConfigModal.vue:42]: Add a warning icon.

File: src/components/DateTimeConfigModal.vue [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+  },
+  mounted(){
+    this.dateTimeFormat = this.currentDateTimeFormat
+    this.currentDate = DateTime.now().toFormat(this.dateTimeFormat)
+  },
+  methods: {
+    closeModal() {
+      modalController.dismiss({ dismissed: true });
+    },
+    async saveAlert() {
```
Comment (disha1202) [inline src/components/DateTimeConfigModal.vue:105]: Improve method name.

File: src/views/OrderDetail.vue [hunk 6]
```diff
@@ -122,9 +122,9 @@
               <ion-label>{{ item.quantityOrdered }} {{ $t("Ordered") }}</ion-label>
             </ion-chip>
 
-            <ion-chip outline class="tablet">
+            <ion-chip outline class="tablet" @click="configureFormat(item.arrivalDate)">
```
Comment (disha1202) [inline src/views/OrderDetail.vue:125]: Improve method name.

File: src/views/PurchaseOrder.vue [hunk 7]
```diff
@@ -173,7 +173,7 @@ export default defineComponent({
             return {
               orderId: item[this.fieldMapping.orderId],
               shopifyProductSKU: item[this.fieldMapping.productSku],
-              arrivalDate: DateTime.fromFormat(item[this.fieldMapping.orderDate], this.dateTimeFormat).toFormat(this.dateTimeFormat), //This is to verify whether the date format is correct.
+              arrivalDate: item[this.fieldMapping.orderDate], //This is to verify whether the date format is correct.
```
Comment (disha1202) [inline src/views/PurchaseOrder.vue:176]: Please remove the comment as it is not valid anymore.

**Review by disha1202 — CHANGES_REQUESTED** [review]

File: src/components/DateTimeConfigModal.vue [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+          
+    <ion-item>
+      <ion-label>
+        <p>{{ $t("Parsed output") }}</p>
+      </ion-label>
+      <ion-label slot="end">{{ isDateInvalid(arrivalDate) ? $t("Invalid input") : arrivalDate }}</ion-label>
+    </ion-item>
+
+    <ion-item lines="none">
+      <ion-label>
```
Comment (disha1202) [inline src/components/DateTimeConfigModal.vue:46]: Use ion-note tag

**Review by disha1202 — CHANGES_REQUESTED** [review]

File: src/components/DateTimeConfigModal.vue (hunk 1 above)
Comment (k2maan) [reply src/components/DateTimeConfigModal.vue]: Updated

File: src/components/DateTimeConfigModal.vue (hunk 8 above)
Comment (k2maan) [reply src/components/DateTimeConfigModal.vue]: Done with the changes.

File: src/views/OrderDetail.vue [hunk 9]
```diff
@@ -401,6 +402,18 @@ export default defineComponent({
         })
         this.isParentProductUpdated = false;
       }
+    },
+    async configureDateTimeFormat(arrivalDate: string) {
+      const timeZoneModal = await modalController.create({
```
Comment (disha1202) [inline src/views/OrderDetail.vue:407]: Improve the variable name.

File: src/components/DateTimeConfigModal.vue [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+  IonNote,
+  IonTitle,
+  IonToolbar,
+  modalController,
+  alertController } from "@ionic/vue";
+import { defineComponent } from "vue";
+import { close, saveOutline } from "ionicons/icons";
+import { mapGetters } from 'vuex';
+import { DateTime } from 'luxon';
+import { warningOutline } from 'ionicons/icons'
```
Comment (disha1202) [inline src/components/DateTimeConfigModal.vue:69]: Use single import for icons.

**Review by disha1202 — CHANGES_REQUESTED** [review]

File: src/views/OrderDetail.vue [hunk 11]
```diff
@@ -402,6 +403,18 @@ export default defineComponent({
         })
         this.isParentProductUpdated = false;
       }
+    },
+    async configureDateTimeFormat(arrivalDate: string) {
+      const dateTimeConfigModal = await modalController.create({
+        component: DateTimeConfigModal,
```
Comment (adityasharma7) [inline src/views/OrderDetail.vue:409]: Improve name, we are not configuring the date time, but the format
```suggestion
        component: ConfigureDateTimeFormatModal,
```

**Review by adityasharma7 — CHANGES_REQUESTED** [review]

File: src/components/ConfigureDateTimeFormatModal.vue [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+      <ion-buttons slot="start">
+        <ion-button @click="closeModal"> 
+          <ion-icon :icon="close" />
+        </ion-button>
+      </ion-buttons>
+      <ion-title>{{ $t("Configure date time format") }}</ion-title>
+    </ion-toolbar>
+  </ion-header>
+
+  <ion-content class="ion-padding">
```
Comment (disha1202) [inline src/components/ConfigureDateTimeFormatModal.vue:13]: Remove class ion-padding

File: src/components/ConfigureDateTimeFormatModal.vue [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+      <a href="https://moment.github.io/luxon/#/formatting?id=table-of-tokens" target="_blank">{{ $t('View all date time formats supported by the HotWax Import app.') }}</a>
+    </ion-item>
+          
+    <ion-item class="ion-padding-bottom">
+      <ion-label color="medium">{{ $t("Parsed output") }}</ion-label>
+      <ion-label slot="end">{{ isDateInvalid(arrivalDate) ? $t("Invalid input") : arrivalDate }}</ion-label>
+      <ion-icon slot="end" color="danger" v-if="isDateInvalid(arrivalDate)" :icon="warningOutline"></ion-icon>
+    </ion-item>
+
+    <ion-note class="ion-text-wrap ion-padding">{{ $t("Changes to the date time format will cause edits done to your PO to be reverted.") }}</ion-note>
```
Comment (disha1202) [inline src/components/ConfigureDateTimeFormatModal.vue:38]: Use ion-label inside ion-item instead od ion-note and remove class ion-padding

File: src/components/ConfigureDateTimeFormatModal.vue [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+    <ion-item>
+      <ion-label color="medium">{{ $t("Expected input") }}</ion-label>
+      <ion-label slot="end">{{ currentDate }}</ion-label>
+    </ion-item>
+    <ion-item lines="full">
+      <ion-label>{{ $t("Custom date format") }}</ion-label>
+      <ion-input clear-input='true' @keyup.enter="dateTimeFormat = $event.target.value;" v-model="dateTimeFormat" :value="dateTimeFormat" :placeholder="defaultDateTimeFormat" />
+    </ion-item>
+
+    <ion-item class="ion-padding-bottom" lines="none">
```
Comment (disha1202) [inline src/components/ConfigureDateTimeFormatModal.vue:28]: Remove class ion-padding-bottom

File: src/components/ConfigureDateTimeFormatModal.vue [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+      <a href="https://moment.github.io/luxon/#/formatting?id=table-of-tokens" target="_blank">{{ $t('View all date time formats supported by the HotWax Import app.') }}</a>
+    </ion-item>
+          
+    <ion-item class="ion-padding-bottom">
```
Comment (disha1202) [inline src/components/ConfigureDateTimeFormatModal.vue:32]: Use class ion-margin-top instead of ion-padding-bottom

**Review by disha1202 — CHANGES_REQUESTED** [review]

File: src/components/ConfigureDateTimeFormatModal.vue (hunk 15 above)
Comment (k2maan) [reply src/components/ConfigureDateTimeFormatModal.vue]: Updated

File: src/components/ConfigureDateTimeFormatModal.vue (hunk 14 above)
Comment (k2maan) [reply src/components/ConfigureDateTimeFormatModal.vue]: Updated

File: src/components/ConfigureDateTimeFormatModal.vue (hunk 13 above)
Comment (k2maan) [reply src/components/ConfigureDateTimeFormatModal.vue]: Updated

**Review by disha1202 — APPROVED** [review]

**Review by adityasharma7 — APPROVED** [review]

**Pull request merged**